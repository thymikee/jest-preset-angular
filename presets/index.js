const snapshotSerializers = require('../build/serializers');

const basePreset = {
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
      stringifyContentPathRegex: '\\.html$',
    },
  },
  testEnvironment: 'jsdom',
  transform: {
    '^.+\\.(ts|js|html)$': 'jest-preset-angular',
  },
  moduleFileExtensions: ['ts', 'html', 'js', 'json'],
  snapshotSerializers,
};

module.exports = {
  defaults: basePreset,
  defaultsESM: {
    ...basePreset,
    extensionsToTreatAsEsm: ['.ts'],
    globals: {
      'ts-jest': {
        ...basePreset.globals['ts-jest'],
        useESM: true,
      },
    },
    moduleNameMapper: {
      tslib: 'tslib/tslib.es6.js',
    },
    transformIgnorePatterns: ['node_modules/(?!tslib)'],
  },
};
