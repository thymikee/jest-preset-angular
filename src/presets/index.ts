import type { JestConfigWithTsJest, TsJestTransformerOptions } from 'ts-jest';

import snapshotSerializers from '../serializers';

const baseConfig: Pick<JestConfigWithTsJest, 'testEnvironment' | 'moduleFileExtensions' | 'snapshotSerializers'> = {
  testEnvironment: 'jsdom',
  moduleFileExtensions: ['ts', 'html', 'js', 'json', 'mjs'],
  snapshotSerializers,
};

const defaultTransformerOptions: TsJestTransformerOptions = {
  tsconfig: '<rootDir>/tsconfig.spec.json',
  stringifyContentPathRegex: '\\.(html|svg)$',
};

const defaultPreset = {
  ...baseConfig,
  transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
  transform: {
    '^.+\\.(ts|js|mjs|html|svg)$': ['jest-preset-angular', defaultTransformerOptions],
  },
};

const defaultEsmPreset = {
  ...baseConfig,
  extensionsToTreatAsEsm: ['.ts'],
  moduleNameMapper: {
    tslib: 'tslib/tslib.es6.js',
  },
  transform: {
    '^.+\\.(ts|js|html|svg)$': [
      'jest-preset-angular',
      {
        ...defaultTransformerOptions,
        useESM: true,
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!tslib)'],
};

export { defaultPreset, defaultEsmPreset, defaultTransformerOptions };
