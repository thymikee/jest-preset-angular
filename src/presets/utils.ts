import type { Config } from 'jest';

import snapshotSerializers from '../serializers';

type BasePresetConfig = Required<Pick<Config, 'testEnvironment' | 'moduleFileExtensions' | 'snapshotSerializers'>>;

export const basePresetConfig: BasePresetConfig = {
    testEnvironment: 'jsdom',
    moduleFileExtensions: ['ts', 'html', 'js', 'json', 'mjs'],
    snapshotSerializers,
};
