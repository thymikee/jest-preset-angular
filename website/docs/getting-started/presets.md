---
id: presets
title: Presets
---

In Jest, **presets** are pre-defined configurations that help streamline and standardize the process of setting up testing environments.
They allow developers to quickly configure Jest with specific transformers, file extensions, and other options.

`jest-preset-angular` provides very opinionated presets and based on what we found to be useful.

:::important

The current best practice for using presets is to call one of the utility functions below to create (and optionally extend) presets. Legacy presets are listed at the bottom of the page.

:::

## Functions

import TOCInline from '@theme/TOCInline';

<TOCInline toc={toc.slice(1)} />

---

### `createCjsPreset(options)`

Create a configuration to process JavaScript/TypeScript/HTML/SVG files (`ts|js|mjs|html|svg`).

#### Parameters

- `options` (**OPTIONAL**)
  - `tsconfig`: see more at [tsconfig options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/tsconfig)
  - `isolatedModules`: see more at [isolatedModules options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/isolatedModules)
  - `astTransformers`: see more at [astTransformers options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/astTransformers)
  - `diagnostics`: see more at [diagnostics options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/diagnostics)

#### Returns

An object contains Jest config:

```ts
type CjsPresetTransformerOptions = {
  tsconfig: string;
  stringifyContentPathRegex: string;
};

type CjsPresetType = {
  testEnvironment: string;
  moduleFileExtensions: Array<string>;
  snapshotSerializers: Array<string>;
  transformIgnorePatterns: Array<string>;
  transform: {
    '^.+\\.(ts|js|mjs|html|svg)$': ['jest-preset-angular', CjsPresetTransformerOptions];
  };
};
```

#### Example:

```ts title="jest.config.ts"
import { createCjsPreset } from 'jest-preset-angular/presets';
import type { Config } from 'jest';

const presetConfig = createCjsPreset({
  //...options
});

export default {
  ...presetConfig,
} satisfies Config;
```

### `createEsmPreset(options)`

Create a configuration to process JavaScript/TypeScript/HTML/SVG files (`ts|js|html|svg`).

#### Parameters

- `options` (**OPTIONAL**)
  - `tsconfig`: see more at [tsconfig options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/tsconfig)
  - `isolatedModules`: see more at [isolatedModules options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/isolatedModules)
  - `astTransformers`: see more at [astTransformers options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/astTransformers)
  - `diagnostics`: see more at [diagnostics options page](https://kulshekhar.github.io/ts-jest/docs/getting-started/options/diagnostics)

#### Returns

An object contains Jest config:

```ts
type EsmPresetTransformerOptions = {
  tsconfig: string;
  stringifyContentPathRegex: string;
  useEsm: true;
};

type EsmPresetType = {
  testEnvironment: string;
  moduleFileExtensions: Array<string>;
  snapshotSerializers: Array<string>;
  extensionsToTreatAsEsm: Array<string>;
  transformIgnorePatterns: Array<string>;
  transform: {
    '^.+\\.(ts|js|html|svg)$': ['jest-preset-angular', EsmPresetTransformerOptions];
  };
};
```

#### Example:

```ts title="jest.config.ts"
import { createEsmPreset } from 'jest-preset-angular/presets';
import type { Config } from 'jest';

const presetConfig = createEsmPreset({
  //...options
});

export default {
  ...presetConfig,
} satisfies Config;
```
