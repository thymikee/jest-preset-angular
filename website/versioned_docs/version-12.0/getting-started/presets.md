---
id: presets
title: Presets
---

### The presets

`jest-preset-angular` comes with 2 presets, covering most of the project's base configuration:

| Preset name                                                        | Description                                                                                                                       |
| ------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------- |
| `jest-preset-angular/presets/default`<br/>or `jest-preset-angular` | TypeScript, JavaScript and HTML files (`js`, `.ts`, `.html`) will be transformed by `jest-preset-angular` to **CommonJS** syntax. |
| `jest-preset-angular/presets/defaults-esm`<br/>                    | TypeScript, JavaScript and HTML files (`js`, `.ts`, `.html`) will be transformed by `jest-preset-angular` to **ESM** syntax.      |

### Basic usage

In most cases, simply setting the `preset` key to the desired preset name in your Jest config should be enough to start
using TypeScript with Jest (assuming you added `jest-preset-angular` to your `devDependencies` of course):

```js
// jest.config.js
module.exports = {
  // [...]
  // Replace `jest-preset-angular` with the preset you want to use
  // from the above list
  preset: 'jest-preset-angular',
};
```

```json
// OR package.json
{
  // [...]
  "jest": {
    // Replace `jest-preset-angular` with the preset you want to use
    // from the above list
    "preset": "jest-preset-angular"
  }
}
```

### Advanced

All presets come with default `ts-jest` config options.
If you want to override any of the options, you'll need to use the JavaScript version of Jest config,
copy the original options and override the options you need:

:::important

If you choose to override `globals` in order to point at a specific tsconfig, you will need to make sure that original `ts-jest`
options provided through the default preset are defined to the `globals.ts-jest` section too, otherwise you will get
errors.

:::

```js
// jest.config.js
const { defaults: jestNgPreset } = require('jest-preset-angular/presets');
// const { defaultsESM: jestNgPreset } = require('jest-preset-angular/presets')

module.exports = {
  // [...]
  globals: {
    'ts-jest': {
      ...jestNgPreset.globals['ts-jest'],
      // [...your overriden options]
    },
  },
};
```

Or through TypeScript (if `ts-node` is installed):

```ts
// jest.config.ts
import type { InitialOptionsTsJest } from 'ts-jest/dist/types';
import { defaults as jestNgPreset } from 'jest-preset-angular/presets';
// import { defaultsESM as jestNgPreset } from 'jest-preset-angular/presets'

const config: InitialOptionsTsJest = {
  // [...]
  globals: {
    'ts-jest': {
      ...jestNgPreset.globals['ts-jest'],
      // [...your overriden options]
    },
  },
};

export default config;
```
