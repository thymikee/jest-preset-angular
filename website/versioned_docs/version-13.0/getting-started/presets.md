---
id: presets
title: Presets
---

### The presets

`jest-preset-angular` comes with 2 presets, covering most of the project's base configuration:

| Preset name                                                        | Description                                                                                                                       |
| ------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------- |
| `jest-preset-angular/presets/default`<br/>or `jest-preset-angular` | TypeScript, JavaScript and HTML files (`js`, `.ts`, `.html`) will be transformed by `jest-preset-angular` to **CommonJS** syntax. |
| `jest-preset-angular/presets/defaults-esm`<br/>                    | TypeScript, JavaScript and HTML files (`js`, `.ts`, `.html`) will be transformed by `jest-preset-angular` to **ESM** syntax.      |

### Basic usage

In most cases, simply setting the `preset` key to the desired preset name in your Jest config should be enough to start
using TypeScript with Jest (assuming you added `jest-preset-angular` to your `devDependencies` of course):

```js tab
module.exports = {
  // [...]
  // Replace `jest-preset-angular` with the preset you want to use
  // from the above list
  preset: 'jest-preset-angular',
};
```

```ts tab
import type { Config } from 'jest';

const jestConfig: Config = {
  // [...]
  // Replace `jest-preset-angular` with the preset you want to use
  // from the above list
  preset: 'jest-preset-angular',
};

export default jestConfig;
```

```JSON tab
{
  //...
  "jest": {
    // Replace `jest-preset-angular` with the preset you want to use
    // from the above list
    "preset": "jest-preset-angular"
  }
}
```

### Advanced

All presets come with default `ts-jest` config options.
If you want to override any of the options, you'll need to use the JavaScript version of Jest config,
copy the original options and override the options you need:

:::important

If you choose to override `transform` in order to point at a specific tsconfig, you will need to make sure that original `ts-jest`
options provided through the default preset are defined to the `transform` section too, otherwise you will get
errors.

:::

```js tab
const { defaultTransformerOptions } = require('jest-preset-angular/presets');
// const { defaultTransformerOptions } = require('jest-preset-angular/presets')

module.exports = {
  // [...]
  transform: {
    '^.+\\.(ts|js|mjs|html|svg)$': [
      'jest-preset-angular',
      {
        ...defaultTransformerOptions,
        // [...your overriden options]
      },
    ],
  },
};
```

```ts tab
import type { Config } from 'jest';
import presets from 'jest-preset-angular/presets';

const jestConfig: Config = {
  // [...]
  transform: {
    '^.+\\.(ts|js|mjs|html|svg)$': [
      'jest-preset-angular',
      {
        ...presets.defaultTransformerOptions,
        // [...your overriden options]
      },
    ],
  },
};

export default jestConfig;
```
