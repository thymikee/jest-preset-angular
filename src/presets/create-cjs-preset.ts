import type { Config } from 'jest';
import type { TsJestTransformerOptions } from 'ts-jest';

import { basePresetConfig, type JSDOMEnvironment } from './utils';

type CjsPresetType = typeof basePresetConfig & Required<Pick<Config, 'transformIgnorePatterns' | 'transform'>>;

type CjsPresetOptionsType = {
    tsconfig?: TsJestTransformerOptions['tsconfig'];
    astTransformers?: TsJestTransformerOptions['astTransformers'];
    babelConfig?: TsJestTransformerOptions['babelConfig'];
    diagnostics?: TsJestTransformerOptions['diagnostics'];
    testEnvironment?: JSDOMEnvironment;
};

export const createCjsPreset = (options: CjsPresetOptionsType = {}): CjsPresetType => {
    const { testEnvironment, ...rest } = options;

    return {
        ...basePresetConfig,
        testEnvironment: testEnvironment ?? basePresetConfig.testEnvironment,
        transformIgnorePatterns: ['node_modules/(?!(.*\\.mjs$|@angular/common/locales/.*\\.js$))'],
        transform: {
            '^.+\\.(ts|js|mjs|html|svg)$': [
                'jest-preset-angular',
                {
                    tsconfig: '<rootDir>/tsconfig.spec.json',
                    stringifyContentPathRegex: '\\.(html|svg)$',
                    ...rest,
                },
            ],
        },
    };
};
