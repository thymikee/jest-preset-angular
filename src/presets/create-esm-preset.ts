import type { Config } from 'jest';
import type { TsJestTransformerOptions } from 'ts-jest';

import { basePresetConfig, type JSDOMEnvironment } from './utils';

type EsmPresetType = typeof basePresetConfig &
    Required<Pick<Config, 'extensionsToTreatAsEsm' | 'moduleNameMapper' | 'transformIgnorePatterns' | 'transform'>>;

type EsmPresetOptionsType = {
    tsconfig?: TsJestTransformerOptions['tsconfig'];
    astTransformers?: TsJestTransformerOptions['astTransformers'];
    babelConfig?: TsJestTransformerOptions['babelConfig'];
    diagnostics?: TsJestTransformerOptions['diagnostics'];
    testEnvironment?: JSDOMEnvironment;
};

export const createEsmPreset = (options: EsmPresetOptionsType = {}): EsmPresetType => {
    const { testEnvironment, ...rest } = options;

    return {
        ...basePresetConfig,
        testEnvironment: testEnvironment ?? basePresetConfig.testEnvironment,
        extensionsToTreatAsEsm: ['.ts'],
        moduleNameMapper: {
            tslib: 'tslib/tslib.es6.js',
        },
        transformIgnorePatterns: ['node_modules/(?!tslib)'],
        transform: {
            '^.+\\.(ts|js|html|svg)$': [
                'jest-preset-angular',
                {
                    tsconfig: '<rootDir>/tsconfig.spec.json',
                    stringifyContentPathRegex: '\\.(html|svg)$',
                    useESM: true,
                    ...rest,
                },
            ],
        },
    };
};
