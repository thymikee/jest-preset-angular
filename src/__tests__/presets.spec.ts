import presets from '../../presets';

describe('Presets', () => {
  test('should return the correct default preset value', () => {
    expect(presets.defaults).toMatchSnapshot('default-preset');
  });

  test('should return the correct ESM preset value', () => {
    expect(presets.defaultsESM).toMatchSnapshot('esm-preset');
  });
});
