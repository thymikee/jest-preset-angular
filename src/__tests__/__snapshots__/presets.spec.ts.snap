// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Presets should return the correct ESM preset value: esm-preset 1`] = `
Object {
  "extensionsToTreatAsEsm": Array [
    ".ts",
  ],
  "globals": Object {
    "ts-jest": Object {
      "stringifyContentPathRegex": "\\\\.html$",
      "tsconfig": "<rootDir>/tsconfig.spec.json",
      "useESM": true,
    },
  },
  "moduleFileExtensions": Array [
    "ts",
    "html",
    "js",
    "json",
  ],
  "moduleNameMapper": Object {
    "^app/(.*)$": "<rootDir>/src/app/$1",
    "^assets/(.*)$": "<rootDir>/src/assets/$1",
    "^environments/(.*)$": "<rootDir>/src/environments/$1",
    "^src/(.*)$": "<rootDir>/src/$1",
    "tslib": "<rootDir>/node_modules/tslib/tslib.es6.js",
  },
  "snapshotSerializers": Array [
    "jest-preset-angular/build/serializers/html-comment",
    "jest-preset-angular/build/serializers/ng-snapshot",
    "jest-preset-angular/build/serializers/no-ng-attributes",
  ],
  "testEnvironment": "jsdom",
  "transform": Object {
    "^.+\\\\.(ts|js|html)$": "jest-preset-angular",
  },
  "transformIgnorePatterns": Array [
    "node_modules/(?!tslib)",
  ],
}
`;

exports[`Presets should return the correct default preset value: default-preset 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "stringifyContentPathRegex": "\\\\.html$",
      "tsconfig": "<rootDir>/tsconfig.spec.json",
    },
  },
  "moduleFileExtensions": Array [
    "ts",
    "html",
    "js",
    "json",
  ],
  "moduleNameMapper": Object {
    "^app/(.*)$": "<rootDir>/src/app/$1",
    "^assets/(.*)$": "<rootDir>/src/assets/$1",
    "^environments/(.*)$": "<rootDir>/src/environments/$1",
    "^src/(.*)$": "<rootDir>/src/$1",
  },
  "snapshotSerializers": Array [
    "jest-preset-angular/build/serializers/html-comment",
    "jest-preset-angular/build/serializers/ng-snapshot",
    "jest-preset-angular/build/serializers/no-ng-attributes",
  ],
  "testEnvironment": "jsdom",
  "transform": Object {
    "^.+\\\\.(ts|js|html)$": "jest-preset-angular",
  },
}
`;
